NIFI-296: Extend capability of IdentifyMimeType

adamonduty opened the pull request:
```
This commit backs IdentifyMimeType with the Apache Tika library. Tika
provides detailed mime type identification such as the ability to
differentiate normal zip files from OOXML MS Office documents.

The mime.type attribute continues to be set, though some mime types
have changed due to Tika naming them differently. In addition,
the mime.extension attribute is set to provide the commonly used
extension for the mime type (if known).
```

Some additional notes about this commit:

I removed the IDENTIFY_ZIP and IDENTIFY_TAR properties. Keeping IDENTIFY_ZIP doesn't make sense because Tika is designed to identify container formats like zip files. Excluding zip files from detection would exclude a number of common mime types, which seems like undesirable behavior. IDENTIFY_TAR is in a similar situation.

Also, in both cases, the previous code would "identify" a zip or tar file by attempting to open them with Zip and Tar readers. I believe Tika will use magic byte detection as a filtering mechanism to avoid applying deep inspection logic (ie opening the zip with a reader) when not necessary.

It takes about 2 seconds to bring up the Tika detectors, which makes the tests run longer, but I believe the detection itself is roughly in the same performance category. The code shares a Tika config and list of detectors to minimize the performance impact related to bringing up detectors.

I also replaced the test resource `1.tar` with a version created by a modern version of tar. The previous tar didn't use the <a href="http://en.wikipedia.org/wiki/Tar_%28computing%29#UStar_format">ustar format</a>, which was standardized in 1988. Tika also couldn't identify the previous tar using magic byte
detection.

And finally, a few of the detected mime types changed names due to Tika naming them differently.

markap14 commented on nifi/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/IdentifyMimeType.java:
```diff
@@ -239,87 +128,39 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+                        mimetype = config.getMimeRepository().forName(mediatype.toString());
+                        extensionRef.set(mimetype.getExtension());
+                    } catch (MimeTypeException ex) {
+                        logger.warn("MIME type detection failed: {}", new Object[]{ex.toString()});
```
I would use "new Object[] {ex}" rather than "new Object[] {ex.toString()}"... this way, if the user configures logback to use debug-level logging, the logger will automatically get the stack trace.

adamonduty replied:
Didn't know that! I'll fix and re-push.

markap14 commented on nifi/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/IdentifyMimeType.java:
```diff
@@ -327,148 +168,41 @@ public void process(final InputStream in) throws IOException {
-                return false;
+        public MediaType detect(InputStream in, Metadata mtdt) throws IOException {
+            // Sanity check the stream. This may not be a tarfile at all
+            in.mark(FlowFilePackagerV1.FILENAME_ATTRIBUTES.length());
```
I'm not sure that I follow the logic here. It appears that you are expecting the first several bytes to be the filename of the attributes file... this is not going to happen for a .tar file, as the first several bytes will be tar header info.
We should probably wait until after Tika has identified a file as a .tar file and then if it is a .tar file perform the logic below to check the filename of the first file. If it's a match then call it flowfile-stream-v1, else call it tar.

adamonduty replied:
The beauty of this approach is the first bytes of the <a href="http://en.wikipedia.org/wiki/Tar_%28computing%29#File_header">tar header</a> are the name of the file. This assumes the first file will be `flowfile.attributes`, but the later code assumes the same. Also, its important to avoid invoking the tar reader in this situation, because this detector runs on every file (the custom detectors run before Tika's detectors).

adamonduty replied:
For example,

```
$ xxd flowfilev1.tar
0000000: 666c 6f77 6669 6c65 2e61 7474 7269 6275  flowfile.attribu
0000010: 7465 7300 0000 0000 0000 0000 0000 0000  tes.............
```

adamonduty replied:
And at the risk of being too chatty, I'll admit that I like your idea of hooking into Tika when it identifies the tar, but I did not see a straightforward way to do so with the Tika API. At least in the class hierarchy for <a href="http://tika.apache.org/1.7/api/">1.7</a>, there is no mention of tar at all, which leaves me to believe that tar is done entirely by magic byte detection, and I'm unsure if Tika offers any flexible way of hooking into that process.

markap14 replied:
Don't worry about being too chatty -- better to be chatty and get this right than to be quiet and provide an inferior product :)

markap14 replied:
wow good call on the first 100 bytes being the filename -- I looked up the tar format to see if that was indeed the case but found this big, verbose, confusing explanation of the header that I didn't understand -- should have tried wikipedia first. I am sorry that I doubted you :)

markap14 commented on nifi/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/IdentifyMimeType.java:
```diff
@@ -327,148 +168,41 @@ public void process(final InputStream in) throws IOException {
+            // Sanity check the stream. This may not be a tarfile at all
+            in.mark(FlowFilePackagerV1.FILENAME_ATTRIBUTES.length());
+            byte[] bytes = new byte[FlowFilePackagerV1.FILENAME_ATTRIBUTES.length()];
+            in.read(bytes);
```
I would use "StreamUtils.fillBuffer(in, bytes, false)" here, to make sure that we get at least the number of bytes we need. I can't imagine that a call to read() in this case would read fewer than the 19 bytes that we need... but just to be sure :)

adamonduty commented:
Pushed a fix to both of those issues.

adamonduty commented:
I agree - that is too much bloat. We could depend on tika-core only, but that provides much more basic functionality when it comes to container detection. Let me investigate a tika-core only solution and some alternatives and get back to you.

adamonduty commented:
The latest commits only depend upon tika-core, which brings the nifi-standard-nar to 13 MB (instead of 37). 

I added sample OOXML office docs from <a href="http://digitalcorpora.org/corpora/govdocs">govdocs</a>, which should be redistributable. 

I also added tests for the mime.extension property.

markap14 commented:
Adam,

Excellent job on this! I sent an e-mail to the digitalcorpora people via their 'contact us' page to ask for info on their license. I can't find anything listed. They indicate that it's free to download and distribute but it could be licensed under GPL or something like that, which is not OK. But if they indicate that it's a compatible license, then I will merge this into develop.

This is a lot simpler than my terrible implementation and likely a lot more powerful!

Many thanks for putting in all the hard work to get this going!

-Mark

adamonduty commented:
Thanks for merges! All of the documents in govdocs were obtained from .gov websites, so I imagine they're in the public domain and fully redistributable. If not, feel free to create some new docs and drop them in (I'll admit being too lazy to find a copy of MS Office myself).
